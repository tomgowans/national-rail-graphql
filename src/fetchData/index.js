const parseString = require("xml2js").parseString;
const fetch = require("node-fetch");

const fetchData = ({ body, tokenValue }) => {
  return new Promise((resolve, reject) => {
    if (!body || !tokenValue) {
      reject(new Error("Error with credentials"));
    } else {
      fetch("https://lite.realtime.nationalrail.co.uk/OpenLDBWS/ldb11.asmx", {
        method: "POST",
        headers: {
          "Content-Type": "text/xml"
        },
        body: `<soap:Envelope xmlns:soap="http://www.w3.org/2003/05/soap-envelope" xmlns:typ="http://thalesgroup.com/RTTI/2013-11-28/Token/types" xmlns:ldb="http://thalesgroup.com/RTTI/2017-10-01/ldb/">
                  <soap:Header>
                    <typ:AccessToken>
                      <typ:TokenValue>${tokenValue}</typ:TokenValue>
                    </typ:AccessToken>
                  </soap:Header>
                  <soap:Body>
                    ${body}
                  </soap:Body>
                </soap:Envelope>`
      })
        .then(response => response.text())
        .then(response => response.replace(/(<\/?)(\w*:)/gm, "$1"))
        .then(response => {
          parseString(response, (error, result) => {
            if (error) {
              return reject(error);
            } else {
              return resolve(result.Envelope.Body[0]);
            }
          });
        })
        .catch(err => {
          // console.error("Error returned from fetch", err);
          return reject(err);
        });
    }
  });
};

module.exports = fetchData;
