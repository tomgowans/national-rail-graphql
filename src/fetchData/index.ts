const parseString = require("xml2js").parseString;
import fetch from "node-fetch";
import { Result } from "../types";

type Response = {
  body: string;
  tokenValue: string;
};

const fetchData = ({ body, tokenValue }: Response): Promise<Result> => {
  return new Promise((resolve, reject) => {
    if (!body || !tokenValue) {
      reject(new Error("Error with credentials"));
    } else {
      fetch("https://lite.realtime.nationalrail.co.uk/OpenLDBWS/ldb11.asmx", {
        method: "POST",
        headers: {
          "Content-Type": "text/xml"
        },
        body: `<soap:Envelope xmlns:soap="http://www.w3.org/2003/05/soap-envelope" xmlns:typ="http://thalesgroup.com/RTTI/2013-11-28/Token/types" xmlns:ldb="http://thalesgroup.com/RTTI/2017-10-01/ldb/">
                  <soap:Header>
                    <typ:AccessToken>
                      <typ:TokenValue>${tokenValue}</typ:TokenValue>
                    </typ:AccessToken>
                  </soap:Header>
                  <soap:Body>
                    ${body}
                  </soap:Body>
                </soap:Envelope>`
      })
        .then(response => response.text())
        .then(response => response.replace(/(<\/?)(\w*:)/gm, "$1"))
        .then(response => {
          parseString(response, (error: Error, result: any) => {
            if (error) {
              return reject(error);
            } else {
              return resolve(result.Envelope.Body[0]);
            }
          });
        })
        .catch((err: Error) => {
          // console.error("Error returned from fetch", err);
          return reject(err);
        });
    }
  });
};

export default fetchData;
