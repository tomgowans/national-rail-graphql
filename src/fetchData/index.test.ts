import fetchData from './index';
import responseData from './response';

jest.mock('node-fetch');

// eslint-disable-next-line @typescript-eslint/no-var-requires
const fetch = require('node-fetch');
const { Response, Headers } = jest.requireActual('node-fetch');

describe('fetchData', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should fetch the data correctly', async () => {
    fetch.mockResolvedValueOnce(
      Promise.resolve(
        new Response(responseData.text, {
          status: 200,
          statusText: 'fail',
          headers: new Headers({
            'Content-Type': 'text/xml',
          }),
        }),
      ),
    );

    const result = await fetchData({
      body: 'body items',
      tokenValue: 'ABC-123',
    });
    expect(result).toMatchSnapshot();
    expect(fetch).toBeCalledTimes(1);
  });

  it('should fail with incorrect credentials', async () => {
    expect.assertions(1);

    await expect(
      fetchData({
        body: 'body items',
        tokenValue: '',
      }),
    ).rejects.toThrow('Error with credentials');
  });

  it('should fail with a malformed XML response', async () => {
    fetch.mockResolvedValueOnce(
      Promise.resolve(
        new Response(
          '<xml><malformedxml/>',

          {
            status: 200,
            statusText: 'fail',
            headers: new Headers({
              'Content-Type': 'text/xml',
            }),
          },
        ),
      ),
    );

    const result = fetchData({
      body: 'RespondWithXMLError',
      tokenValue: 'ABC-123',
    });

    await expect(result).rejects.toThrow('Unclosed root tag\nLine: 0\nColumn: 20\nChar: ');
  });
});

// TODO: Server error response?
