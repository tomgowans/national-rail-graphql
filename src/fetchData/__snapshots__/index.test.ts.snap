// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchData should fetch the data correctly 1`] = `
Object {
  "GetDepartureBoardResponse": Array [
    Object {
      "$": Object {
        "xmlns": "http://thalesgroup.com/RTTI/2017-10-01/ldb/",
      },
      "GetStationBoardResult": Array [
        Object {
          "$": Object {
            "xmlns:lt": "http://thalesgroup.com/RTTI/2012-01-13/ldb/types",
            "xmlns:lt2": "http://thalesgroup.com/RTTI/2014-02-20/ldb/types",
            "xmlns:lt3": "http://thalesgroup.com/RTTI/2015-05-14/ldb/types",
            "xmlns:lt4": "http://thalesgroup.com/RTTI/2015-11-27/ldb/types",
            "xmlns:lt5": "http://thalesgroup.com/RTTI/2016-02-16/ldb/types",
            "xmlns:lt6": "http://thalesgroup.com/RTTI/2017-02-02/ldb/types",
            "xmlns:lt7": "http://thalesgroup.com/RTTI/2017-10-01/ldb/types",
          },
          "crs": Array [
            "ECR",
          ],
          "generatedAt": Array [
            "2019-11-20T21:09:59.5101527+00:00",
          ],
          "locationName": Array [
            "East Croydon",
          ],
          "platformAvailable": Array [
            "true",
          ],
          "trainServices": Array [
            Object {
              "service": Array [
                Object {
                  "destination": Array [
                    Object {
                      "location": Array [
                        Object {
                          "crs": Array [
                            "GTW",
                          ],
                          "locationName": Array [
                            "Gatwick Airport",
                          ],
                        },
                      ],
                    },
                  ],
                  "etd": Array [
                    "On time",
                  ],
                  "length": Array [
                    "12",
                  ],
                  "operator": Array [
                    "Thameslink",
                  ],
                  "operatorCode": Array [
                    "TL",
                  ],
                  "origin": Array [
                    Object {
                      "location": Array [
                        Object {
                          "crs": Array [
                            "BDM",
                          ],
                          "locationName": Array [
                            "Bedford",
                          ],
                        },
                      ],
                    },
                  ],
                  "platform": Array [
                    "6",
                  ],
                  "serviceID": Array [
                    "8NvDNS9Gpylp0MOg/MmCxw==",
                  ],
                  "serviceType": Array [
                    "train",
                  ],
                  "std": Array [
                    "21:07",
                  ],
                },
                Object {
                  "destination": Array [
                    Object {
                      "location": Array [
                        Object {
                          "crs": Array [
                            "TBD",
                          ],
                          "locationName": Array [
                            "Three Bridges",
                          ],
                        },
                      ],
                    },
                  ],
                  "etd": Array [
                    "On time",
                  ],
                  "operator": Array [
                    "Thameslink",
                  ],
                  "operatorCode": Array [
                    "TL",
                  ],
                  "origin": Array [
                    Object {
                      "location": Array [
                        Object {
                          "crs": Array [
                            "LBG",
                          ],
                          "locationName": Array [
                            "London Bridge",
                          ],
                        },
                      ],
                    },
                  ],
                  "platform": Array [
                    "3",
                  ],
                  "serviceID": Array [
                    "rkEPUnFHelVWfAllpJoHZA==",
                  ],
                  "serviceType": Array [
                    "train",
                  ],
                  "std": Array [
                    "21:08",
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
  ],
}
`;
